src/uu/env/src/split_iterator.rs

```diff
@@ -0,0 +1,505 @@
+// This file is part of the uutils coreutils package.
+//
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+//
+// This file is based on work from Tomasz Miąsko who published it as "shell_words" crate,
+// licensed under the Apache License, Version 2.0 <LICENSE-APACHE>
+// or the MIT license <LICENSE-MIT>, at your option.
+//
+//! Process command line according to parsing rules of original GNU env.
+//! Even though it looks quite like a POSIX syntax, the original
+//! "shell_words" implementation had to be adapted significantly.
+//!
+//! Apart from the grammar differences, there is a new feature integrated: $VARIABLE expansion.
+//!
+//! [GNU env] <https://www.gnu.org/software/coreutils/manual/html_node/env-invocation.html#g_t_002dS_002f_002d_002dsplit_002dstring-syntax>
+// spell-checker:ignore (words) Tomasz Miąsko rntfv FFFD varname
+
+#![forbid(unsafe_code)]
+
+use std::mem;
+use std::ops::Range;
+
+use crate::parse_error::ParseError;
+use crate::raw_string_parser::is_ascii;
+use crate::raw_string_parser::RawStringExpander;
+use crate::raw_string_parser::RawStringParser;
+
+#[derive(Clone, Copy)]
+pub enum State {
+    /// Within a delimiter.
+    Delimiter,
+    /// After backslash, but before starting word.
+    DelimiterBackslash,
+    /// Within an unquoted word.
+    Unquoted,
+    /// After backslash in an unquoted word.
+    UnquotedBackslash,
+    /// Within a single quoted word.
+    SingleQuoted,
+    /// After backslash inside a double quoted word.
+    SingleQuotedBackslash,
+    /// Within a double quoted word.
+    DoubleQuoted,
+    /// After backslash inside a double quoted word.
+    DoubleQuotedBackslash,
+    /// Inside a comment.
+    Comment,
+}
+
+const BACKSLASH: u8 = b'\\';
+const DOUBLE_QUOTES: u8 = b'\"';
+const SINGLE_QUOTES: u8 = b'\'';
+const INVALID_UTF8_MARKER: char = '\u{FFFD}';
+
+const REPLACEMENTS: (&[u8], &[u8]) = ("rntfv_#$\"".as_bytes(), "\r\n\t\x0C\x0B #$\"".as_bytes());
```

Comment:
I think this'll be easier as an `&[(u8, u8)]`, which will also allow us to remove the `static_assertions` crate.

Reply:
done